pkg/trace/{api,config}: allow read & write timeouts to be configured separately

cimi opened the pull request:
### What does this PR do?

This PR allows the read and write timeouts for the HTTP server in the APM trace-agent 
to be configured separately. Eventually, we may want to increase the default write timeout,
but this PR only changes the configuration mechanism, it does not change any of the 
existing defaults.

### Motivation

Customers have reported that they see warnings for profile upload
failures in their application logs, coming from `dd-trace-java`.
However, the data is uploaded successfully and is available in the
Datadog app.

We see a similar behaviour internally with services configured to
send profiling data to multiple targets, we see failure warnings
in the logs, but the data is uploaded successfully and available
in the UI.

The cause of the problem is the timeout changed by this PR - whenever
the profile upload request takes longer than the 5s timeout we have
configured at present, the http server in `datadog-agent` gracefully
closes the connection, `dd-trace-java` sees an empty reply from
server and triggers the UnexpectedEndOfStream exception which is logged.

This is more likely to happen for customers that have large profiles or
slow network connections to our intake.

We've changed the logging in dd-trace-java to use a friendlier message
(e.g. 'WARN ProfileUploader - No reply from server, please check upload
status' instead of 'Failed to upload') and we will not log the stacktrace
for these exceptions as they're not useful.

However, since for some customers this can be logged every five minutes,
ideally we would eliminate the error altogether. This change allows configuring 
read and write timeouts separately so we can increase the default write 
timeout in the future.

### Additional Notes

Alternatively, we could return a 200 OK from the agent immediately after we
receive the profiling data from the application, so the client app logs will
not see any errors. However, even though errors would be available in the
datadog-agent logs, I'm not sure customers would look there so this could mask
real upload problems (bad api key, bad config etc).

### Describe your test plan

We have automated tests for the bits of configuration changed in this PR.

Commit:
Configure read/write timeouts separately in trace-agent

Customers have reported that they see warnings for profile upload
failures in their application logs, coming from `dd-trace-java`.
However, the data is uploaded successfully and is available in the
Datadog app.

We see a similar behaviour internally with services configured to
send profiling data to multiple targets, we see failure warnings
in the logs, but the data is uploaded successfully and available
in the UI.

The cause of the problem is the timeout configured on the http server
in the trace-agent which receives profiling and trace data. When the
profile upload request takes longer than the 5s timeout we have
configured at present, the http server in `datadog-agent` gracefully
closes the connection, `dd-trace-java` sees an empty reply from
server and triggers the UnexpectedEndOfStream exception which is logged.

This is more likely to happen for customers that have large profiles or
slow network connections to our intake.

This change allows configuring the read and write timeouts separately.
The read timeout doesn't affect this issue since datadog-agent doesn't
have any problem consuming data from the client.

cimi commented:
Related PR: https://github.com/DataDog/dd-trace-java/pull/2196

AlexJF commented:
> Alternatively, we could return a 200 OK from the agent immediately after we
receive the profiling data from the application, so the client app logs will
not see any errors. However, even though errors would be available in the
datadog-agent logs, I'm not sure customers would look there so this could mask
real upload problems (bad api key, bad config etc).

👍 on this not being a very good option due to the hiding of problems.

Found https://github.com/golang/go/issues/16100 talking about per-handler timeouts and streaming-related timeout handling in Go which seems what we'd want in this situation. Looks like the read/write timeouts in go are for the entire request whereas in other frameworks they are usually timeouts between individual reads/writes in a request. While we don't have that I guess bumping this timeout to 16 seconds doesn't sound that bad. For APM IIRC there are 2 flows: trace and stats data:
* Stats - smaller payloads, send fast
* Traces - potentially bigger payloads if there are a lot of spans. I think intake has a 30MB limit as well here so it's a very similar situation to profiling and it's likely clients with big traces are already facing issues from missing responses.

It would change a bit the behaviour around connection loss/issues, potentially requiring 10 extra seconds of queueing but this sounds reasonable for APM data? If there's a separate connection timeout I'd probably keep that at a lower value (e.g. 16 for read/write but 5 for connect?).

Commit:
Increase trace-agent http server write timeout to 15s

This allows us to keep the connection open for longer than the previous
default value of 5s, so we avoid the connection being terminated during
profile upload requests to the intake that take longer than that.

Commit:
Add debug logging for profiling intake URL configuration

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -130,14 +130,18 @@ func (r *HTTPReceiver) buildMux() *http.ServeMux {
-	timeout := 5 * time.Second
-	if r.conf.ReceiverTimeout > 0 {
-		timeout = time.Duration(r.conf.ReceiverTimeout) * time.Second
+	readTimeout := 5 * time.Second
```
Let's move these to `pkg/trace/config.New()` as defaults. It makes more sense.

gbbr commented on pkg/trace/api/api_test.go:
```diff
@@ -954,6 +954,23 @@ func BenchmarkWatchdog(b *testing.B) {
+func TestReceiverTimeoutConfig(t *testing.T) {
+	tcs := [][]int{{1, 2, 1, 2}, {1, 0, 1, 15}, {0, 1, 5, 1}, {0, 0, 5, 15}}
+	for _, tc := range tcs {
+		t.Run("receiver timeout", func(t *testing.T) {
```
```suggestion
		t.Run("", func(t *testing.T) {
```
In Go it's usually just one word for the identifier. In this case, if you leave it blank, Go will just add numbers for each test, like an index, which is good enough here.

gbbr commented on pkg/trace/api/api_test.go:
```diff
@@ -954,6 +954,23 @@ func BenchmarkWatchdog(b *testing.B) {
 }
 
+func TestReceiverTimeoutConfig(t *testing.T) {
+	tcs := [][]int{{1, 2, 1, 2}, {1, 0, 1, 15}, {0, 1, 5, 1}, {0, 0, 5, 15}}
```
Can we move these inline? e.g.
```go
for _, tc := range [][]int{
    {1, 2, 1, 2},
    {1, 0, 1, 15},
    ...
} {
    ...
}
```

gbbr commented on pkg/trace/api/profiles.go:
```diff
@@ -57,6 +57,9 @@ func profilingEndpoints(apiKey string) (urls []*url.URL, apiKeys []string, err e
 		}
 	}
+	for i, u := range urls {
+		log.Debug("Set profiling intake URL [%d]:%s", i, u)
```
```suggestion
		log.Debugf("Set profiling intake URL [%d]: %s", i, u)
```
I'd expect the linter to catch this.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -348,7 +348,14 @@ func (c *AgentConfig) loadDeprecatedValues() error {
+		c.ReceiverReadTimeout = cfg.GetInt("apm_config.receiver_timeout")
+		c.ReceiverWriteTimeout = cfg.GetInt("apm_config.receiver_timeout")
+	}
+	if cfg.IsSet("apm_config.receiver_read_timeout") {
```
You'll need to add some extra lines around here: https://github.com/DataDog/datadog-agent/blob/05e067be4fac699854971b6ca8d58cc724fd30ee/pkg/config/apm.go#L57

Something like:
```go
	config.BindEnv("apm_config.receiver_write_timeout", "DD_APM_RECEIVER_WRITE_TIMEOUT")                             //nolint:errcheck
	config.BindEnv("apm_config.receiver_read_timeout", "DD_APM_RECEIVER_READ_TIMEOUT")                             //nolint:errcheck

```

Commit:
Address review: code style and move defaults to constructor

Commit:
Add env var bindings for receiver timeouts

Commit:
Add release note for APM timeout config update

Commit:
Revert write timeout to the previous default of 5s

We want to try out this configuration internally before setting it
as a default for all users.

gbbr commented on pkg/trace/config/config.go:
```diff
@@ -135,6 +137,9 @@ func New() *AgentConfig {
 		ReceiverPort:    8126,
 		MaxRequestBytes: 50 * 1024 * 1024, // 50MB
 
+		ReceiverReadTimeout:  time.Duration(5) * time.Second,
+		ReceiverWriteTimeout: time.Duration(5) * time.Second,
```
I think the case here is not needed
```suggestion
		ReceiverReadTimeout:  5 * time.Second,
		ReceiverWriteTimeout: 5 * time.Second,
```

gbbr reviewed (approved):
Thanks! LGTM! If you'd like to merge this feel free to take it out of draft! ⚠️ But careful! We are in freeze so no merging until that's done :)

cimi committed:
Remove unnecessary time.Duration in config defaults

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

hush-hush commented on pkg/trace/api/profiles.go:
```diff
@@ -57,6 +57,9 @@ func profilingEndpoints(apiKey string) (urls []*url.URL, apiKeys []string, err e
 		}
 	}
+	for i, u := range urls {
+		log.Debugf("Set profiling intake URL [%d]:%s", i, u)
```
Are those scrubs ? URLs might contains password that we don't want in the logs and then in the flares.

AlexJF replied:
We send the API key via header in profiling so these URLs should be clean. They have to point directly to profiling intake addresses in different environments (e.g. `intake.profile.datadoghq.com`, `intake.profile.datadoghq.eu`). 

I guess someone could decide to put whatever they wanted on one of those places but if we follow that route then we'd have to potentially scrub all configuration fields.

AlexJF replied:
If we were worried about proxy configurations, these go through the same configuration path as APM and are not logged anywhere: https://github.com/DataDog/datadog-agent/blob/6275ef2adaf6dcacff9a3b0630d2e8ecf7762629/pkg/trace/config/config.go#L235

hush-hush replied:
I'm worried about people using a proxy with a login+password. We do scrub the login+password when logging them everywhere.

AlexJF replied:
Are we talking about using a proxy without actually configuring it as such in the agent but just specifying some weird address in the url? Wasn't aware this was something people did.

But isn't this covered by us using the agent-wide logger which already seems to scrub all logged messages? https://github.com/DataDog/datadog-agent/blob/6c1c5208173f3643b14679964c01a03c2ee2af3f/pkg/util/log/log.go#L179-L180

hush-hush commented on pkg/trace/config/apply.go:
```diff
@@ -348,7 +348,17 @@ func (c *AgentConfig) loadDeprecatedValues() error {
+		c.ReceiverWriteTimeout = d * time.Second
+	}
+	if cfg.IsSet("apm_config.receiver_read_timeout") {
+		d := time.Duration(cfg.GetInt("apm_config.receiver_read_timeout"))
+		c.ReceiverReadTimeout = d * time.Second
```
why not in one line ?

```suggestion
		c.ReceiverReadTimeout  = time.Duration(cfg.GetInt("apm_config.receiver_read_timeout")) * time.Second
```

hush-hush commented on pkg/trace/config/apply.go:
```diff
@@ -348,7 +348,17 @@ func (c *AgentConfig) loadDeprecatedValues() error {
+		c.ReceiverReadTimeout = d * time.Second
+	}
+	if cfg.IsSet("apm_config.receiver_write_timeout") {
+		d := time.Duration(cfg.GetInt("apm_config.receiver_write_timeout"))
+		c.ReceiverWriteTimeout = d * time.Second
```
Same here

hush-hush commented on releasenotes/notes/apm-receiver-timeout-config-6c66ef404b36b766.yaml:
```diff
@@ -0,0 +1,14 @@
+---
+features:
+  - |
+    APM: Allow configuring read and write timeouts for the trace-agent HTTP
+    receiver separately using apm_config.receiver_write_timeout and
+    apm_config.receiver_read_timeout.
```
If we're listing them in the change log we should also add them in the datadog.yaml.example file.

remeh commented on pkg/config/apm.go:
```diff
@@ -55,6 +55,8 @@ func setupAPM(config Config) {
 	config.BindEnvAndSetDefault("apm_config.windows_pipe_security_descriptor", "D:AI(A;;GA;;;WD)", "DD_APM_WINDOWS_PIPE_SECURITY_DESCRIPTOR") //nolint:errcheck
 
 	config.BindEnv("apm_config.receiver_timeout", "DD_APM_RECEIVER_TIMEOUT")                             //nolint:errcheck
```
The `apm_config.receiver_timeout` doesn't seem to be used anymore anywhere in the codebase, should we remove it here?
This raises another question: if already used by someone, isn't that an issue to replace it this way? Do you think the CHANGELOG will be enough? If so, we should probably mention how to migrate from one to the other in the CHANGELOG for instance.

cimi replied:
It's still used here: https://github.com/DataDog/datadog-agent/blob/cimi/PROF-2503-increase-apm-http-server-timeout/pkg/trace/config/apply.go#L350-L354

My intention is to have the two more specific configurations (read and write) override the generic one when set.

gbbr replied:
It is used: https://github.com/DataDog/datadog-agent/pull/6976/files#diff-cf6f1068cf79cfbcdb88f347f0b847a7ea377be2bb432081c6ebdbc7250d9166R351

Commit:
Set receiver timeouts in a single statement

gbbr commented:
@cimi freeze is Friday, does this get pushed to the next release? If not, time is limited :)

cimi commented:
> @cimi freeze is Friday, does this get pushed to the next release? If not, time is limited :)

Sorry for the silence! We've decided to drop this change for now - we've not received any more reports from customers about the timeout issue after we explained what was happening and how they can increase the timeout for both read and write using the environment variable.

The reason why we wanted to split the config was to reduce the risk of changing the default timeout, but it comes with the downside of having more configuration properties, overrides etc. so let's hold off on doing this until we know it's necessary.

Thanks for the help, we can resurrect this code if we decide to do this later 😃 cc @AlexJF
